sabnzbd/downloader.py

```diff
@@ -875,11 +838,49 @@ def __handle_recv_result(self, nw: NewsWrapper, bytes_received: int = 0, done: b
                 nw.article = server.get_article()
                 if nw.article:
                     self.__request_article(nw)
-                    return
+                    return bytes_received
             server.busy_threads.remove(nw)
             server.idle_threads.append(nw)
             self.remove_socket(nw)
 
+        return bytes_received
+
+    def __check_speed(self):
+        if not self.bandwidth_limit:
+            return
+
+        BPSMeter = sabnzbd.BPSMeter
+        if BPSMeter.bps + BPSMeter.sum_cached_amount > self.bandwidth_limit:
+            BPSMeter.update()
+            while BPSMeter.bps > self.bandwidth_limit:
+                time.sleep(0.01)
+                BPSMeter.update()
+
+    def __check_assembler(self):
+        """See if we need to delay because the queues are full"""
+        # Sleep for an increasing amount of time, depending on queue sizes.
+        if (assembler_level := sabnzbd.Assembler.queue_level()) > SOFT_QUEUE_LIMIT:
+            time.sleep(assembler_level - SOFT_QUEUE_LIMIT)
+            sabnzbd.BPSMeter.delayed_assembler += 1
+            logged_counter = 0
+
+            while not self.shutdown and sabnzbd.Assembler.queue_level() >= 1:
+                # Only log/update once every second, to not waste any CPU-cycles
+                if not logged_counter % 10:
+                    # Make sure the BPS-meter is updated
+                    sabnzbd.BPSMeter.update()
+
+                    # Update who is delaying us
+                    logging.debug(
+                        "Delayed - %d seconds - Assembler queue: %d",
+                        logged_counter / 10,
+                        sabnzbd.Assembler.queue.qsize(),
+                    )
+
+                # Wait and update the queue sizes
+                time.sleep(0.1)
+                logged_counter += 1
+
     def __finish_connect_nw(self, nw: NewsWrapper) -> bool:
```

Comment:
I think we will also need locks for error handling, there's all kinds of setting error messages and increasing/decreasing server threads that need to be atomic to work correctly.

Reply:
I've added the things I've identified to handle to the list.

I think some operations are atomic and don't need locks like `server.bad_cons = 0`.

What about the read_fds dictionary?
`self.read_fds[nw.nntp.fileno] = nw`
I could just use synchronized on add_socket and remove_socket but as far as I can tell they are safe without?

Some like `__finish_connect_nw` I'll just stick a synchronized decorator around since they are rarely called anyway.

`article.nzf.nzo.update_download_stats(sabnzbd.BPSMeter.bps, server.id, nw.data_position)`
I think a global lock in downloader or a lock per nzo would probably also work but likely not worth it.